lxd/devlxd.go

```diff
@@ -297,6 +298,64 @@ func devlxdDevicesGetHandler(d *Daemon, c instance.Instance, w http.ResponseWrit
 	return response.DevLxdResponse(http.StatusOK, c.ExpandedDevices(), "json", c.Type() == instancetype.VM)
 }
 
+var devlxdUbuntuProGet = devLxdHandler{
+	path:        "/1.0/ubuntu-pro",
+	handlerFunc: devlxdUbuntuProGetHandler,
+}
+
+func devlxdUbuntuProGetHandler(d *Daemon, c instance.Instance, w http.ResponseWriter, r *http.Request) response.Response {
+	if shared.IsFalse(c.ExpandedConfig()["security.devlxd"]) {
+		return response.DevLxdErrorResponse(api.NewGenericStatusError(http.StatusForbidden), c.Type() == instancetype.VM)
+	}
+
+	if r.Method != http.MethodGet {
+		return response.DevLxdErrorResponse(api.NewGenericStatusError(http.StatusMethodNotAllowed), c.Type() == instancetype.VM)
+	}
+
+	s := d.State()
+	hostAttachSetting := s.UbuntuPro.GuestAttach()
+	guestAttachSetting := c.ExpandedConfig()["ubuntu_pro.guest_attach"]
+
+	// If either the host or the guest have the value set to "off" (or if the attachment is not configured on the instance), then return "off".
+	if hostAttachSetting == pro.GuestAttachSettingOff || guestAttachSetting == string(pro.GuestAttachSettingOff) || guestAttachSetting == "" {
+		return response.DevLxdResponse(http.StatusOK, pro.SettingsFile{GuestAttach: pro.GuestAttachSettingOff}, "json", c.Type() == instancetype.VM)
+	}
+
+	// Otherwise, return the value from the instance configuration.
+	return response.DevLxdResponse(http.StatusOK, pro.SettingsFile{GuestAttach: pro.GuestAttachSetting(guestAttachSetting)}, "json", c.Type() == instancetype.VM)
+}
+
+var devlxdUbuntuProTokenPost = devLxdHandler{
+	path:        "/1.0/ubuntu-pro/token",
+	handlerFunc: devlxdUbuntuProTokenPostHandler,
+}
+
+func devlxdUbuntuProTokenPostHandler(d *Daemon, c instance.Instance, w http.ResponseWriter, r *http.Request) response.Response {
+	if shared.IsFalse(c.ExpandedConfig()["security.devlxd"]) {
+		return response.DevLxdErrorResponse(api.NewGenericStatusError(http.StatusForbidden), c.Type() == instancetype.VM)
+	}
+
+	if r.Method != http.MethodPost {
+		return response.DevLxdErrorResponse(api.NewGenericStatusError(http.StatusMethodNotAllowed), c.Type() == instancetype.VM)
+	}
+
+	// Return http.StatusForbidden if the host does not have guest attachment enabled.
+	s := d.State()
+	hostAttachSetting := s.UbuntuPro.GuestAttach()
+	if hostAttachSetting == pro.GuestAttachSettingOff {
+		return response.DevLxdErrorResponse(api.NewStatusError(http.StatusForbidden, "Guest attachment not allowed"), c.Type() == instancetype.VM)
+	}
+
+	// Run pro guest attach command.
+	stdout, err := shared.RunCommandContext(r.Context(), "pro", "api", "u.pro.attach.guest.get_guest_token.v1")
```

Comment:
Lets push this into the ubuntupro package please and add a function to s.UbuntuPro to get a token.